File: Code/Java/Anagram.java
/*
Two strings are said to be an Anagram if all the characters of a string are equal and have same frequency but arranged in different order.
The simple logic to solve is taking the two strings into a character array and sort both the arrays and again store it to the same string variable,
if both the strings are equal then one string is said to be anagram of another.
*/
package p1;

import java.util.*;
public class Anagram {

	public static void main(String[] args) {
	
		Scanner sc = new Scanner(System.in);
		
		System.out.println("Enter string 1: ");
		String s = sc.next();
		System.out.println("Enter string 2: ");
		String s1 = sc.next();
		 // storing the length of both the strings       
		int l1 = s.length(),l2 = s1.length(); 
		
		if(l1!=l2) {
			//if both are of unequal length then its for sure that they are not an anagram
			System.out.println("Two strings are not Anagram"); 
		}
		else {
			char c1[] = s.toCharArray(),c2[] = s1.toCharArray();
			
			s = ""; s1 = "";
			//sorting both the array.
			Arrays.sort(c1);Arrays.sort(c2);
			
			for(char ch:c1) {
				s = s + ch;
			}
			
			for(char ch:c2) {
				s1 = s1 + ch;
			}
			if(s.compareTo(s1)==0) {
				System.out.println("Two strings are Anagram");
			}else
				System.out.println("Two strings are not Anagram");
				
		}
	}
}


/* 
Test Cases
1.	     
INPUT: 
Enter string 1: prateek
Enter string 2: eektarp
OUTPUT:
Two strings are Anagram

2.
INPUT:
Enter string 1: abcde
Enter string 2: cdba       
OUTPUT:
Two strings are not Anagram
      
      
Time Complexity-O(nlog n)
Space Complexity-O(n) 
where n: number of characters in string
    
    */

File: Code/Python/Anagram.py
"""
Problem Statement: Python program to check given two strings are Anagram of each other or not .
    
    An anagram of a string is an  another string that contains the same characters but the order of characters can be different.
For example ,listen and silent are anagrams of each oter because all the characters in the string "listen" are present in string "silent".
So, listen and silent are anagrams of each other. 
     At First convert both strings to list of characters and sort them using sort() methid and check if both two list strings are equal print 
    strings are anagrams of each other ,else print strings are not anagrams of eachother. 
 """
def anagram(string1,string2):
     list_string1=list(string1)
     list_string1.sort()
     list_string2=list(string2)
     list_string2.sort()
     if(list_string1==list_string2):
         print("Given strings are anagrams of eachother")
     else:
         print("Given strings are not anagrams of eachother")
string1=input("enter string1:")
string2=input("enter string2:")
anagram(string1,string2)

"""
Test cases:
  Input:
  Enter string1: silent
  Enter string2: listen
  Output: Given strings are anagrams of eachother 
Because here after sorting string1 and string2 ,list_string1=['e','i','l','n','s','t'] and list_string2=['e','i','l','n','s','t']. As list_string1 = list_string2 the given 
strings are anagrams of each other.
 
 Input:
 enter string1: hello
 enter string2: world
 Output: Given strings are not anagrams of eachother.
Because here after sorting string1 and string2 ,list_string1=['e','h','l','l','o'] and list_string2=['d','l','o','r','w']. As list_string1 is not equal to list_string2 the given 
strings are  not anagrams of each other.

Time Complexity: O(nlogn) .Because time complexity for sort() function is O(nlogn) where n is the size of input that is the length of string.
Space complexity: O(length(string1+string2))

"""
